### PR TITLE
[Merged by Bors] - feat: port observe tactic from mathlib

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/LibrarySearch.lean [hunk 1]
```diff
@@ -196,3 +196,41 @@ elab tk:"library_search%" : term <= expectedType => do
...
+    let .mvar goal ← mkFreshExprMVar type | failure
+    if let some _ ← librarySearch goal (← librarySearchLemmas.get) [] then
+      reportOutOfHeartbeats tk
+      throwError "observe did not find a solution"
+    else
+      let v := (← instantiateMVars (mkMVar goal)).headBeta
+      if trace.isSome then
+        -- TODO we should be allowed to pass an identifier to `addHaveSuggestion`.
+        addHaveSuggestion tk type v
+      let (_, newGoal) ← (← getMainGoal).let name v
```
Comment (gebner) [inline Mathlib/Tactic/LibrarySearch.lean:228]:
```suggestion
      let (_, newGoal) ← (← getMainGoal).note name v
```

At least that's what the docstring promises.

File: Mathlib/Tactic/LibrarySearch.lean [hunk 2]
```diff
@@ -196,3 +196,41 @@ elab tk:"library_search%" : term <= expectedType => do
...
+
+open Elab.Tactic Elab Tactic in
+elab_rules : tactic |
+  `(tactic| observe%$tk $[?%$trace]? $[$n?:ident]? : $t:term $[using $[$required:term],*]?) => do
+  let mainGoal ← getMainGoal
+  let name : Name := match n? with
+    | none   => `this
+    | some n => n.getId
+  mainGoal.withContext do
+    let type ← elabTermWithHoles t none (← getMainTag) true <&> (·.1)
```
Comment (gebner) [inline Mathlib/Tactic/LibrarySearch.lean:218]:
```suggestion
    let (type, _) ← elabTermWithHoles t none (← getMainTag) true
```

File: Mathlib/Tactic/LibrarySearch.lean [hunk 3]
```diff
@@ -196,3 +196,41 @@ elab tk:"library_search%" : term <= expectedType => do
...
+syntax (name := observe) "observe" "?"? (ident)? ":" term (" using " (colGt term),+)? : tactic
+
+open Elab.Tactic Elab Tactic in
+elab_rules : tactic |
+  `(tactic| observe%$tk $[?%$trace]? $[$n?:ident]? : $t:term $[using $[$required:term],*]?) => do
+  let mainGoal ← getMainGoal
+  let name : Name := match n? with
+    | none   => `this
+    | some n => n.getId
+  mainGoal.withContext do
```
Comment (gebner) [inline Mathlib/Tactic/LibrarySearch.lean:217]:
```suggestion
  withMainContext do
```

**Review by gebner — APPROVED** [review]
bors d+

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
Gah. 

bors merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**